libs/nss.c

```diff
@@ -57,53 +57,80 @@ static FILE *tcb_safe_open(const char *file, const char *name)
 	return f;
 }
 
-int _nss_tcb_getspnam_r(const char *name, struct spwd *__result_buf,
-    char *__buffer, size_t __buflen, struct spwd **__result)
+enum nss_status _nss_tcb_getspnam_r(const char *name,
+    struct spwd *__result_buf, char *__buffer, size_t __buflen, int *__errnop)
 {
 	FILE *f;
 	char *file;
 	int retval, saved_errno;
+	struct spwd **result_buf_ptr = NULL;
 
-	if (asprintf(&file, TCB_FMT, name) < 0)
+	if (asprintf(&file, TCB_FMT, name) < 0) {
+		*__errnop = EAGAIN;
 		return NSS_STATUS_TRYAGAIN;
+	}
+
 	f = tcb_safe_open(file, name);
 	free(file);
-	if (!f)
+	if (!f) {
+		*__errnop = ENOENT;
 		return NSS_STATUS_UNAVAIL;
+	}
 
-	retval = fgetspent_r(f, __result_buf, __buffer, __buflen, __result);
+	retval = fgetspent_r(f, __result_buf, __buffer,
+	    __buflen, result_buf_ptr);
 	saved_errno = errno;
 	fclose(f);
 	errno = saved_errno;
-	if (!retval)
+	if (!retval) {
+		if (!*result_buf_ptr) {
+			*__errnop = ENOENT;
+			return NSS_STATUS_NOTFOUND;
+		}
+
 		return NSS_STATUS_SUCCESS;
+	}
 
 	switch (saved_errno) {
 	case 0:
+		if (!*result_buf_ptr) {
+			*__errnop = ENOENT;
+			return NSS_STATUS_NOTFOUND;
+		}
+
 		return NSS_STATUS_SUCCESS;
 
 	case ENOENT:
+		*__errnop = ENOENT;
 		return NSS_STATUS_NOTFOUND;
 
+	case EAGAIN:
+		*__errnop = EAGAIN;
+		return NSS_STATUS_TRYAGAIN;
+
 	case ERANGE:
+		*__errnop = ERANGE;
 		return NSS_STATUS_TRYAGAIN;
 
 	default:
+		*__errnop = ENOENT;
 		return NSS_STATUS_UNAVAIL;
 	}
 }
 
-int _nss_tcb_getspent_r(struct spwd *__result_buf,
-    char *__buffer, size_t __buflen, struct spwd **__result)
+enum nss_status _nss_tcb_getspent_r(struct spwd *__result_buf,
+    char *__buffer, size_t __buflen, int *__errnop)
 {
 	struct dirent *result;
 	off_t currpos;
 	int retval, saved_errno;
 
 	if (!tcbdir) {
 		tcbdir = opendir(TCB_DIR);
-		if (!tcbdir)
+		if (!tcbdir) {
+			*__errnop = ENOENT;
```

Comment:
Maybe `*__errnop = errno;`.

Reply:
`NSS_STATUS_UNAVAIL` with `ENOENT` means "A necessary input file cannot be found.", which is the case here; tcbdir cannot be opened.